---
title: "Comparison of Differential Expression"
author: "Tolga Tabanli"
date: "2025-01-25"
output: pdf_document
knit: (function(input, encoding) {
  rmarkdown::render(input,
                    encoding = "UTF-8",
                    output_dir = "/home/tolga/Desktop/IDPs_under_Heatshock/Knittings")})
---

```{r setup_comparison, include=FALSE}
# DONE
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/home/tolga/Desktop/IDPs_under_Heatshock/')
```

This chapter deals with how the results of DESeq analysis was visualized
and used to graphically compare the differences between IDP expression
and the overall expression patterns. I have used bar plots to distinguish
the number of up and down-regulated genes, heatmaps for comparison of
expression patterns and clusters in different genotypes and lastly
gene enrichment analysis to find out the functions of the IDPs
implicated in heat shock and possible broken down in knock-down and
knock-out.

## Libraries and Preparation

The used libraries give an overview of this chapter.
IDPs were read from the RDS data.
The relevant DESeq results from DESeq pipeline were read in
and the label "ref" was used for saves and plots.
The results were lastly filtered for significance.

```{r, eval=T, message=F}
library(tidyverse)
library(UpSetR)
library(gprofiler2)
library(ggpubr)
library(janitor)

idps <- readRDS("IDP decisions/commons_modes.Rds")

deseq_results <- readRDS("deseq_reference37.Rds") %>%
  map(as.data.frame)
ref <- 37

deseq_sig <- deseq_results %>%
  map(drop_na) %>%
  map(arrange, padj) %>%
  map(~ dplyr::select(.x, log2FoldChange, padj)) %>%
  map(~ filter(.x, padj < 0.05)) %>%
  map(~ filter(.x, abs(log2FoldChange) > 0.6)) # for lfc filtering
```

## Up or Down?

To be able follow what the ratio of up- and down-regulated genes
under different conditions was, the filtered DESeq results were 
labeled according to direction of regulation:

```{r, eval=T, message=F}
expr_direction <- deseq_sig %>%
  map(~ mutate(.x, direction = case_when(
    log2FoldChange < 0 ~ "Down",
    log2FoldChange > 0 ~ "Up"
  ))) %>%
  map(rownames_to_column) %>%
  map(~ mutate(.x, idp = case_when(
    rowname %in% idps ~ TRUE,
    !(rowname %in% idps) ~ FALSE
  )))
```

To compare IDPs' expression behaviour with the background (whole genome),
I needed to use a function to convert the up/down labels to counts both for
whole genome and for IDPs, a kind of tabularization. In short,
what this does is 1) summarise counts for Up/Downregulated genes,
2) repeat for IDPs, and 3) return a new data frame consisting of these four
groups.

```{r, eval=T}
convert_to_counts <- function(df) {
  n_total <- df %>%
    group_by(direction) %>%
    summarise(count = n(), .groups = "drop") %>%
    mutate(group = paste0("Total ", direction))
  
  n_idp <- df %>%
    filter(idp == TRUE) %>%
    group_by(direction) %>%
    summarise(count = n(), .groups = "drop") %>%
    mutate(group = paste0("IDP ", direction))
  
  bind_rows(n_total, n_idp)
}
```

Then, the data with information on regulation direction was passed through
tabularization function described above and then plotted on a bar plot.

```{r, eval=T, message=F}
expr_direction %>%
  map_dfr(convert_to_counts, .id = "dataset") %>%
  # filter(str_starts(dataset, "Wildtype")) %>% # for creating Wild-type only
  mutate(group = factor(group, levels = c("Total Up", "Total Down",
                                          "IDP Up", "IDP Down"))) %>%
  ggplot(aes(x = group, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ dataset, ncol = 2, scales = "free",
             labeller = labeller(dataset = function(x) {
               gsub("(\\w+?)_(\\d+?)_(\\d+?)", "\\1 \\2 °C, t = \\3", x)
             })) +
  labs(
    title = "Comparison of Total and IDP Regulation",
    x = "Direction of Regulation of Total vs IDP under heat shock",
    caption = paste("Reference is", ref, "°C of corresponding time and genotype."),
    y = "Count",
    fill = "Category"
  ) +
  theme_minimal() +
  theme(axis.text.x = element_blank()) +
  scale_fill_manual(values = c(
    "IDP Up" = "red",
    "Total Up" = "pink",
    "IDP Down" = "blue",
    "Total Down" = "lightblue"
  ))
```

## UpSet

To see if there is an overlap between the significant IDPs from the results of
different conditions, I used UpSet plot. The row names are the gene names.
Text scale parameters indicate the font sizes used in the plot.

```{r, eval=T, message=F}
deseq_sig_idps <- deseq_sig %>%
  map(row.names) %>%
  map(intersect, idps)

upset(fromList(deseq_sig_idps), sets = names(deseq_sig_idps),
      keep.order = T,
      nsets = length(deseq_sig_idps),
      point.size = 3, line.size = 1,
      mainbar.y.label = "Significant IDP Intersection",
      sets.x.label = "Number of IDPs",
      order.by = c("freq"),
      text.scale = c(1.7, 1, 1.3, 1, 1, 1.3))
```

A similar UpSet plot was created for the intersection on only wild-type.
Here is given only the set for the input to upset():

```{r}
updown_wt <- expr_direction %>%
  keep(str_starts(names(.), "Wildtype")) %>%
  map(filter, idp) %>%
  map(~ mutate(.x, direction = factor(.x$direction, levels = c("Up", "Down")))) %>%
  map(~ split(.x$rowname, .x$direction)) %>%
  flatten() %>%
  set_names(c("Wildtype_42_10 Up", "Wildtype_42_10 Down",
              "Wildtype_42_30 Up", "Wildtype_42_30 Down"))
```

## Heatmaps

Heatmaps were used the compare a possible pattern of expression that
was hypothesized to change between all genes and IDPs. The data preparation
step included:

```{r, eval=T, message=F}
library(pheatmap)
library(tidyverse)

# PREPARE
idps <- readRDS("IDP decisions/commons_modes.Rds")
heat_matrix <- data.frame(rowname = idps) # For IDP step
deseq <- readRDS("deseq_reference37.Rds") %>%
  map(data.frame) %>%
  map(rownames_to_column) %>%
  map(select, rowname, log2FoldChange)
```

Next, matrices of log2 fold changes were created.
For IDP case, an empty matrix with row names as IDPs were used as join partner.
As heatmap cannot accept NA values in the matrix, NA-rows in both matrices
were dropped. (The NA values are naturally generated in DESeq in cases such
as zero or low counts, extreme variability or other statistical test mechanisms.)

```{r, eval=T, message=F}
# Heatmap of IDPs
joined_matrix <- deseq %>%
  imap(~ setNames(.x, gsub("log2FoldChange", .y, names(.x)))) %>%
  map(right_join, heat_matrix) %>% # joins here with IDPs
  purrr::reduce(full_join) %>%
  drop_na() %>%
  column_to_rownames() %>%
  as.matrix()

pheatmap(joined_matrix, show_rownames = F,
         main = "Heatmap from IDPs")


# Heatmap of Non-IDPs
joined_matrix <- deseq %>%
  map(filter, !(rowname %in% idps)) %>%
  imap(~ setNames(.x, gsub("log2FoldChange", .y, names(.x)))) %>%
  purrr::reduce(full_join) %>%
  drop_na() %>%
  column_to_rownames() %>%
  as.matrix()

pheatmap(joined_matrix, show_rownames = F,
         main = "Heatmap of Non-IDPs")
```

Additionally, I wanted to compare the "real" IDPs from DisProt, which
included 212 proteins. The reason why these proteins dod not directly
form the basis of the exploration is given in the Discussion of the Report.

```{r}
# Heatmap of DisProt Proteins
## accession mapping
biomart_mapping <- read_tsv("IDP decisions/biomart_mapping.tsv")

## read the disprot proteins
disprot <- read_tsv("DisProt/DisProt release_2024_12.tsv") %>%
  select(acc) %>%
  distinct() %>%
  inner_join(biomart_mapping, by = join_by(acc == uniprotswissprot)) %>%
  pull(ensembl_gene_id)

heat_matrix_disprot <- data.frame(rowname = disprot)

joined_matrix <- deseq %>%
  imap(~ setNames(.x, gsub("log2FoldChange", .y, names(.x)))) %>%
  map(right_join, heat_matrix_disprot) %>%
  purrr::reduce(full_join) %>%
  drop_na() %>%
  column_to_rownames() %>%
  as.matrix()
```

It should be noted that the heatmaps did not take *significance* into account,
meaning that they were not filtered according to p-value or LFC and
the color labels mirror all the LFC values of genes that had a non-NA value
in each condition.
